rewrite-core/src/main/java/org/openrewrite/trait/TraitMatcher.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.trait;
+
+import org.openrewrite.*;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Incubating(since = "8.30.0")
+public interface TraitMatcher<U extends Trait<?>> {
+
+    Optional<U> get(Cursor cursor);
+
+    Stream<U> higher(Cursor cursor);
+
+    Stream<U> lower(Cursor cursor);
```

Comment:
I'd appreciate a comment helping understand if higher is towards the root or away from the root.

Trees in computer science are upside down from trees in nature, right? I always forget this. A doc indicating which is which would help as a reminder.

Reply:
Good point, added more docs in [ed7d875](https://github.com/openrewrite/rewrite/pull/4309/commits/ed7d8758ebbe4c56c225e37e237c6fc9d1ef8452).